### PR TITLE
[Inductor] Honor cumsum dtype in pointless_cumsum pattern

### LINKED ISSUE #189518: Inductor's pointless_cumsum_replacement drops cumsum's explicit dtype argument
### 🐛 Describe the bug

`pointless_cumsum_replacement` (torch/_inductor/fx_passes/post_grad.py) rewrites `cumsum(full(...))` into an `arange`-based expression, but constructs the replacement using the **full tensor's dtype** and ignores the `dtype=` keyword passed to `cumsum`. When the input is created inside the compiled graph with a different dtype than the requested accumulation dtype, the compiled result has the wrong dtype vs eager:

```python
import torch

def f():
    ones = torch.ones((2, 4, 4), dtype=torch.bool)
    return ones.cumsum(1, dtype=torch.bfloat16)

eager = f()
compiled = torch.compile(f)()
print(eager.dtype)     # torch.bfloat16
print(compiled.dtype)  # torch.int64  <- wrong
```

Disabling the pattern matcher (`torch._inductor.config.pattern_matcher = False`) makes compiled output match eager, isolating the rewrite as the cause.

Real-world impact: huggingface/transformers' DETR-family sine position embeddings used exactly this construct, making every model in that family (DETR, Sam2/Sam3, MaskFormer, …) crash under `torch.compile` in fp16/bf16 with `mat1 and mat2 must have the same dtype, but got Float and BFloat16` — see huggingface/transformers#47228 (transformers is working around it in huggingface/transformers#47238, but the pattern is unsound for any code hitting it).

Expected: the replacement should honor `cumsum`'s `dtype` kwarg (cast the arange expression accordingly), or the pattern should not match when an explicit `dtype` differing from the input's promotion dtype is present.

### Versions

Reproduced on torch 2.13.0 (cpu and mps) and per the transformers issue on 2.x nightlies with CUDA (A100). Pattern present at torch/_inductor/fx_passes/post_grad.py:970 in 2.13.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by guan404ming**
## Related Issue

Fixes #189518

## Why

The `pointless_cumsum_replacement` pattern does not capture cumsum's keyword-only `dtype` argument, so the replacement derives the output dtype from `full`'s dtype alone and silently returns the wrong dtype (e.g. `bool_tensor.cumsum(1, dtype=bfloat16)` compiles to int64). This breaks transformers' DETR-family models under fp16/bf16 compile.

## How

- Read the dtype kwarg from the matched node and cast the replacement result
- Cast `fill_value` the way `full()` does for integral dtypes

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/fx_passes/post_grad.py [hunk 1]
```diff
@@ -1016,16 +1016,20 @@ def pointless_cumsum_replacement(match: Match, shape, fill_value, device, dtype,
...
+    out_dtype = match.output_node().kwargs.get("dtype") or dtype
+
     def repl(*shape):
         dim_size = shape[dim]
         idx = torch.arange(1, dim_size + 1, device=device, dtype=dtype)
 
         inter_shape = [1] * len(shape)
         inter_shape[dim] = dim_size
-        return (idx * fill_value).view(inter_shape).expand(shape)
+        return (idx * fill_value).view(inter_shape).expand(shape).to(out_dtype)
```
Comment (aorenste) [inline torch/_inductor/fx_passes/post_grad.py:1038]: [cleanup] `cumsum(dtype=D)` casts its input to `D` *before* accumulating, but this casts only the result (`.to(out_dtype)`) while `fill_value` and the `arange` above still run in the input's `dtype`. For integral `D` the two are equivalent (modular arithmetic), but for floating `D` they diverge. Beyond the already-noted `float->int` case, a lower-precision input feeding a higher-precision accumulator also silently diverges from eager: `full(fp16, 1.0).cumsum(dim, dtype=fp32)` compiles to `[..., 5000, 5000, 5000]` vs eager `[..., 4998, 4999, 5000]` (the arange runs in fp16 and loses precision eager keeps by casting to fp32 first). The reported `bool->bf16` case is unaffected.

Pre-existing (the `arange` dtype is unchanged here) and non-blocking, but since this PR is specifically about honoring cumsum's `dtype`, keying the `fill_value` cast and the arange precision off `out_dtype` instead of the input dtype closes both cases:

```python
integral_out = is_integer_dtype(out_dtype) or is_boolean_dtype(out_dtype)
if integral_out:
    fill_value = int(bool(fill_value)) if is_boolean_dtype(out_dtype) else int(fill_value)
acc_dtype = torch.int64 if integral_out else torch.float64
# arange in acc_dtype, keep the trailing .to(out_dtype)
```

(retaining the existing "integral input with no explicit dtype promotes to int64" rule when computing `out_dtype`.)

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

File: torch/_inductor/fx_passes/post_grad.py (hunk 1 above)
Comment (guan404ming) [reply torch/_inductor/fx_passes/post_grad.py:1038]: Good catch, just fixed. The `fill_value` cast and accumulation dtype are now keyed off `out_dtype` as suggested, and the existing int64 promotion rule is retained. Added tests for both the fp16 -> fp32 and float -> int cases; the fp16 one reproduces your reported divergence without the fix.

**Comment by guan404ming** [comment]
Thanks!

**Comment by frgossen** [comment]
Apologies @guan404ming — I believe this revert was a mis-attribution and this PR should be relanded.

**Comment by guan404ming** [comment]
Never mind @frgossen, I could help open another pr to reland the chnages. Thanks!
